### PR TITLE
test: Add CloudProvisioner unit tests & clean up old test artifacts

### PR DESCRIPTION
**Pull request by edreed**
**What type of PR is this?**

/kind test

**What this PR does / why we need it**:

1. Adds unit test for `CloudProvisioner`.
2. Removes deprecated, obsolete and/or unused functions originally implemented to facilitate v1/v2 side-by-side testing.
3. Fixes bugs in `nextToken` generation for `ListVolumes` & `ListSnapshots`
4. Fixes `nil` reference bugs in `NewAzureDiskSnapshot` when fields are missing.

**Which issue(s) this PR fixes**:

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: pkg/provisioner/cloudprovisioner.go [hunk 1]
```diff
@@ -585,7 +585,8 @@ func (c *CloudProvisioner) ListSnapshots(
 	// 4. StartingToken is not null, and MaxEntries is not null. Return `MaxEntries` snapshots from `StartingToken`.
 	start := 0
 	if startingToken != "" {
-		start, err := strconv.Atoi(startingToken)
+		var err error
```
Comment (mukhoakash) [inline pkg/provisioner/cloudprovisioner.go:588]: Any particular reason for changing the declaration?

File: pkg/provisioner/cloudprovisioner.go [hunk 2]
```diff
@@ -604,7 +605,8 @@ func (c *CloudProvisioner) ListSnapshots(
 		totalEntries = int(maxEntries)
 	}
 	entries := []v1alpha1.Snapshot{}
-	for count := 0; start < len(snapshots) && count < totalEntries; start++ {
+	count := 0
+	for ; start < len(snapshots) && count < totalEntries; start++ {
```
Comment (mukhoakash) [inline pkg/provisioner/cloudprovisioner.go:609]: This declaration change too?

File: pkg/provisioner/cloudprovisioner.go (hunk 1 above)
Comment (edreed) [reply pkg/provisioner/cloudprovisioner.go:588]: Yes. Using `:=` redeclared another instance of the `start` variable within the `if` statement scope. Any updates within this scope would be lost, and so `ListSnapshots` would always return results from the beginning.

File: pkg/provisioner/cloudprovisioner.go (hunk 2 above)
Comment (edreed) [reply pkg/provisioner/cloudprovisioner.go:609]: Oh, that was an artifact from an earlier attempt to fix the snapshot list `nextToken` calculation by adding the calculated `count` value to `start` on line 622 below. What I didn't realize at the time was that `start` was being incremented by the `for` loop. I reverted the addition, but forgot to revert this declaration. Will fix.

Comment (edreed) [reply pkg/provisioner/cloudprovisioner.go]: Fixed.

**Pull request merged**